src/core/libmaven/PeakDetector.cpp

```diff
@@ -376,103 +527,435 @@ void PeakDetector::processSlices(vector<mzSlice*> &slices, string setName)
+                container.back().setSlice(slice);
+
+                size_t totalSize = container.size();
+                container.back().setGroupId(totalSize);
+                metaGroups[totalSize - 1] = {};
+                for (auto child : orphans)
+                    metaGroups[totalSize - 1].push_back(child);
+            }
 
-                matchFound = true;
-                toBeMerged.push_back(groupWithTarget);
+            return metaGroups;
+        };
+
+    unordered_map<Compound*, unordered_map<size_t, vector<size_t>>> metaGroups;
+
+    // find isotope meta-groups
+    for (auto& elem : nonParentIsotopologues) {
+        if (_mavenParameters->stop)
+            return;
+
+        Compound* compound = elem.first;
+        auto& isotopeIndexes = elem.second;
+        auto metaIsotopeGroups = makeMeta(compound,
+                                          isotopeIndexes,
+                                          parentCompounds,
+                                          [](PeakGroup* group) {
+                                              return group->isotope().name;
+                                          });
+        metaGroups[compound] = metaIsotopeGroups;
+    }
+
+    // find adduct meta-groups
+    for (auto& elem : nonParentAdducts) {
+        if (_mavenParameters->stop)
+            return;
+
+        Compound* compound = elem.first;
+        auto& adductIndexes = elem.second;
+        auto metaAdductGroups = makeMeta(compound,
+                                         adductIndexes,
+                                         parentCompounds,
+                                         [](PeakGroup* group) {
+                                             return group->adduct()->getName();
+                                         });
+        if (metaGroups.count(compound) > 0) {
+            auto& existingMetaGroups = metaGroups[compound];
+            for (auto& elem : metaAdductGroups) {
+                size_t parentIndex = elem.first;
+                vector<size_t>& childIndexes = elem.second;
+                if (existingMetaGroups.count(parentIndex) > 0) {
+                    vector<size_t>& existingChildIndexes =
+                        existingMetaGroups[parentIndex];
+                    existingChildIndexes.insert(end(existingChildIndexes),
+                                                begin(childIndexes),
+                                                end(childIndexes));
+                } else {
+                    existingMetaGroups[parentIndex] = childIndexes;
+                }
             }
+        } else {
+            metaGroups[compound] = metaAdductGroups;
         }
+    }
 
-        if (matchFound) {
-            iter = mavenParameters->allgroups.erase(iter);
-        } else {
-            ++iter;
+    // perform final meta-grouping and queue children to be erased
+    vector<size_t> indexesToErase;
+    for (auto& elem : metaGroups) {
```

Comment:
so metaGroups is "compound_name" : [ parentIndex1:[childIndices], parentIndex2:[childIndices],...parentIndexN:[childIndices] ]

Changing `elem` to `compound_name` might make this easier to comprehend.